crates/circuit/src/dag_node.rs

```diff
@@ -76,114 +109,121 @@ pub struct DAGOpNode {
-                        let key = PyString::new_bound(py, index_strs.join(",").as_str());
-                        cache.set_item(&cache_key, &key)?;
-                        key.into_any()
+        op: Bound<PyAny>,
+        qargs: Option<TupleLikeArg>,
+        cargs: Option<TupleLikeArg>,
+        #[allow(unused_variables)] dag: Option<Bound<PyAny>>,
+    ) -> PyResult<Py<Self>> {
+        let py_op = op.extract::<OperationFromPython>()?;
+        let qargs = qargs.map_or_else(|| PyTuple::empty_bound(py), |q| q.value);
+        let sort_key = qargs.str().unwrap().into();
+        let cargs = cargs.map_or_else(|| PyTuple::empty_bound(py), |c| c.value);
+        let instruction = CircuitInstruction {
+            operation: py_op.operation,
+            qubits: qargs.unbind(),
+            clbits: cargs.unbind(),
+            params: py_op.params,
+            extra_attrs: py_op.extra_attrs,
+            #[cfg(feature = "cache_pygates")]
+            py_op: RefCell::new(Some(op.unbind())),
+        };
+
+        Py::new(
+            py,
+            (
+                DAGOpNode {
+                    instruction,
+                    sort_key,
+                },
+                DAGNode { node: None },
+            ),
+        )
+    }
+
+    fn __hash__(slf: PyRef<'_, Self>) -> PyResult<u64> {
+        let super_ = slf.as_ref();
+        let mut hasher = AHasher::default();
+        hasher.write_isize(super_.py_nid());
+        hasher.write(slf.instruction.operation.name().as_bytes());
+        Ok(hasher.finish())
+    }
+
+    fn __eq__(slf: PyRef<Self>, py: Python, other: &Bound<PyAny>) -> PyResult<bool> {
+        let Ok(other) = other.downcast::<Self>() else {
+            return Ok(false);
+        };
+        let borrowed_other = other.borrow();
+        let other_super = borrowed_other.as_ref();
+        let super_ = slf.as_ref();
+
+        if super_.py_nid() != other_super.py_nid() {
+            return Ok(false);
+        }
+        if !slf
+            .instruction
+            .operation
+            .py_eq(py, &borrowed_other.instruction.operation)?
+        {
+            return Ok(false);
+        }
+        let params_eq = if slf.instruction.operation.try_standard_gate().is_some() {
+            slf.instruction
+                .params
+                .iter()
+                .zip(borrowed_other.instruction.params.iter())
+                .all(|(a, b)| match [a, b] {
+                    [Param::Float(float_a), Param::Float(float_b)] => {
+                        relative_eq!(float_a, float_b, max_relative = 1e-10)
                     }
-                }
-            }
-            None => qargs.str()?.into_any(),
+                    [Param::ParameterExpression(param_a), Param::ParameterExpression(param_b)] => {
+                        param_a.bind(py).eq(param_b).unwrap()
+                    }
+                    _ => false,
+                })
```

Comment:
I did this for semantic equality but I think we're ok here though because the `DAGOpNode.__eq__` method here is really supposed to be used in place of the object identity and set/dict membership which is no longer reliable with this PR. I opted to keep this stricter so we're not accidentally treating equivalent op nodes as the same.

Reply:
Ok, I'm fine with that in principle.  Can you add a comment explaining it just above?

I think we still might need to accept `[Param::Obj(a), Param::Obj(b)] => a.bind(py).eq(b)` in that form though.

I don't feel great about suppressing the exceptions in the `ParameterExpression` case (nor the proposed `Obj` case) - can we propagate that up?